src/System.Management.Automation/engine/PSVersionInfo.cs

```diff
@@ -332,13 +334,22 @@ public override ICollection Keys
         {
             get
             {
-                Array arr = new string[base.Keys.Count];
-                base.Keys.CopyTo(arr, 0);
-                Array.Sort(arr, StringComparer.OrdinalIgnoreCase);
-                return arr;
+                ArrayList keyList = new ArrayList(base.Keys);
+                keyList.Sort();
+                MoveItemOnFirstPlace(PSVersionInfo.PSEditionName, keyList);
+                MoveItemOnFirstPlace(PSVersionInfo.PSVersionName, keyList);
+                return keyList.ToArray();
```

Comment:
This implementation is a little bit heavy, it involves moving keyList's elements around for 3 times (`sort()` and 2 `MoveItemOnFirstPlace`) + 2 collection allocation (`ArrayList` and `ToArray`).
Maybe the following pseudo code would be better?
```CSharp
Array arr = new string[base.Keys.Count];
int index = 2;
foreach (string key in base.Keys)
{
    switch (key)
       case PSVersionInfo.PSVersionName:
                arr[0] = key;
                break;
       case PSVersionInfo.PSEditionName:
                arr[1] = key;
                break;
       default:
                arr[index++] = key;
                break;
}
Array.Sort(arr, 2, arr.Length - 2; /* the equalityComparer passed in the constructor */);
return arr;
```
It allocate only one collection, and elements in `arr` only get moved around once, within the `Sort` method.

The pseudo code is still not perfect, because we will create an array object every time `Keys` is accessed. If we want to go further to avoid that, we can create our own type that implements `ICollection`and is immutable from outside (no public members to mutate it, just like the type of object that `Hashtable.Keys` returns -- `Hashtable+KeyCollection`). Then we can keep one instance of it with `PSVersionHashTable`, and always reuse and return the same instance from `Keys` property.